File: go/fixture/go-mini/internal/mocks/repo.go
// Package mocks holds hand-written doubles for the repository and notifier.
package mocks

import (
	"context"
	"sort"
	"sync"

	"example.com/go-mini/internal/models"
	"example.com/go-mini/internal/repository"
)

// for testing
type Repo struct {
	mu      sync.Mutex
	Devices map[string]models.Device
	Calls   map[string]int
	SaveErr error
}

var _ repository.DeviceRepository = (*Repo)(nil)

// Get gets a device.
func (r *Repo) Get(_ context.Context, tenant, id string) (models.Device, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.count("Get")
	d, ok := r.Devices[tenant+"/"+id]
	if !ok {
		return models.Device{}, repository.ErrNotFound
	}
	return d, nil
}

// Save saves a device.
func (r *Repo) Save(_ context.Context, d models.Device) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.count("Save")
	if r.SaveErr != nil {
		return r.SaveErr
	}
	if r.Devices == nil {
		r.Devices = map[string]models.Device{}
	}
	r.Devices[d.Tenant+"/"+d.ID] = d
	return nil
}

// List lists all devices.
func (r *Repo) List(_ context.Context) ([]models.Device, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.count("List")
	out := make([]models.Device, 0, len(r.Devices))
	for _, d := range r.Devices {
		out = append(out, d)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out, nil
}

func (r *Repo) count(method string) {
	if r.Calls == nil {
		r.Calls = map[string]int{}
	}
	r.Calls[method]++
}

File: py/fixture/py-mini/internal/mocks/repo.py
"""A repository double that counts its calls."""

import threading

from internal.models.device import Device
from internal.repository.device_repository import DeviceRepository
from internal.repository.store import NotFoundError


# for testing
class Repo:  # noqa: D101
    def __init__(self, devices: dict[str, Device] | None = None) -> None:
        self.mu = threading.Lock()
        self.devices: dict[str, Device] = devices if devices is not None else {}
        self.calls: dict[str, int] = {}
        self.save_err: Exception | None = None

    def get(self, tenant: str, device_id: str) -> Device:
        """Get a device."""
        with self.mu:
            self._count("get")
            d = self.devices.get(tenant + "/" + device_id)
        if d is None:
            raise NotFoundError(f"device {tenant}/{device_id} not found")
        return d

    def save(self, d: Device) -> None:
        """Save a device."""
        with self.mu:
            self._count("save")
            if self.save_err is not None:
                raise self.save_err
            self.devices[d.tenant + "/" + d.id] = d

    def list_devices(self) -> list[Device]:
        """List all devices."""
        with self.mu:
            self._count("list")
            return sorted(self.devices.values(), key=lambda d: d.id)

    def _count(self, method: str) -> None:
        self.calls[method] = self.calls.get(method, 0) + 1


_double: type[DeviceRepository] = Repo
